### ISSUE TITLE
Campaign completion markers are not synced via Steam cloud

### ISSUE THREAD

**Issue by bobeff** [wesnoth/wesnoth#4163]
I installed the game via Steam and completed the first few campaigns on all difficulties. I was relaying on Steam cloud service to keep all my save games and game progress, but when I tried the game on the new clean Windows 10 installation, with a huge discontent, I realized that only the saved games have been kept. All markers for completed campaigns on the different difficulty levels are lost. Is it possible in the future versions to sync those markers alongside saved games data via Stream cloud?

**Comment by CelticMinstrel** [comment]
Short answer: It's not synced due to a limitation of the way Steam cloud sync is set up (at least for games not using the Steamworks API), so the chance of changing it to be synced is fairly low. Read on for details.

First of all, the settings Steam offers for cloud sync only lets the developer select a single directory to be synced to the cloud.

Secondly, Wesnoth's userdata directory is poorly structured for the way Steam cloud works. Everything that the game saves is grouped together in a single directory (called the userdata directory), but that includes things that should not be synced to the Steam cloud. For example, all your downloaded add-ons are stored in the userdata directory, and the file that stores saved login info is structured in a way that is not guaranteed to work on another computer, so syncing it would be a waste of time.

On the other hand, not all the data that _could_ be synced is gathered under a single subdirectory of the userdata directory. In addition to saved games, there is the folder where Wesnoth by default saves any maps you create in the map editor, the directory which stores arbitrary global persistent data set by add-ons, and the preferences folder. There might even be something else that I'm missing.

Campaign completion markers are saved in the preferences file, which as mentioned is not synced, so unfortunately they are among the things that have fallen victim to the clash between Steam's design and Wesnoth's folder structure.

So, long story short: unless someone decides to rearrange the structure of the Wesnoth userdata directory, it's unlikely that anything besides saved games will ever be synced; and if someone _did_ decide to rearrange it, I think it might get messy ensuring that past synced data is not lost.

If you do still have access to the old computer, you should be able to get back the progress markers by copying over the preferences file, for the record... but I imagine not everyone in this situation will still have access to the old computer.

**Comment by gfgtdf** [comment]
If someone wants to fix this: note that the preferences file iirc also contains data that is specific to the machine and that one usually would not want to be synced, I'm in particular taking about the last used window size, so to implement this one would also need to split that file.

**Comment by Wedge009** [comment]
Hopping from installation to installation, host to host, if I've wanted to move my campaign progress then yes, it's a matter of manually updating the preferences file on the new installation/host. It's a trivial, but manual, process - it doesn't surprise me that Wesnoth's configuration/progress/etc doesn't carry very well in Steam given that it was only added with release 1.14. It's a nice-to-have, certainly.

**Comment by bobeff** [comment]
Thank you for the feedback. For know I can manually restore my progress markers, but it seems to me that it is not very difficult a new version of the game, just to check for progress markers in the old location and to save them alongside with all the new game progress in separate file in the same directory where the saved games are. It can also clear the old markers after this.

**Comment by Wedge009** [comment]
I've recently noticed for Windows preferences are kept separate between different major version, eg 1.14, 1.15, 1.17. But for Linux - at least my distribution - I only have a single preferences file in `~/.config/wesnoth/` while everything else seems to be in its own major version directory, eg `~/.local/share/wesnoth/1.14/`. So there will be cross-platform complexities to negotiate as well (for the moment I'm still using both Windows and Linux).

**Comment by Wedge009** [comment]
> ...it seems to me that it is not very difficult a new version of the game, just to check for progress markers in the old location and to save them alongside with all the new game progress in separate file in the same directory where the saved games are.

I suppose this is part of why #2710 remains open?

**Comment by CelticMinstrel** [comment]
I don't think that's related to this.

**Comment by Wedge009** [comment]
My remark about the preferences file being shared in Linux is obsolete with #8059 / Wesnoth 1.19.
